### PR TITLE
Pull out target generation from the analyzer

### PR DESCRIPTION
**Pull request by amorde**
## Changes

* Pull out the handling of target generation into its own class
* Fix / add lots of type annotations

This exposed all of the variables needed to generate the targets, and will make it easier to decompose it further down the road. Currently the target generation doesn't have too much direct test coverage, and instead is tested at a higher level. I would write more direct unit tests if I even knew what the expected behavior was based on all the inputs :p

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* Pull out target generation from the analyzer  
  [amorde](https://github.com/amorde)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/cocoapods/installer/analyzer/target_generator.rb [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+            dedupe_cache = {}
+            pod_targets = resolver_specs_by_target.flat_map do |target_definition, specs|
+              grouped_specs = specs.group_by(&:root).values.uniq
+              grouped_specs.flat_map do |pod_specs|
+                generate_pod_target([target_definition], target_inspections, pod_specs.map(&:spec), sandbox, installation_options).scoped(dedupe_cache)
+              end
+            end
+          end
+
+          all_resolver_specs = resolver_specs_by_target.values.flatten.map(&:spec).uniq
```
Comment (amorde) [inline lib/cocoapods/installer/analyzer/target_generator.rb]: The code starting from here was duplicated within each branch of the `if`, with the only difference being:

```
if installation_options.integrate_targets?
  ...
  # within target.test_dependent_targets_by_spec_name =
  hash[test_spec.name] = filter_dependencies(test_dependencies, pod_targets_by_name, target
else
  ...
 # within target.test_dependent_targets_by_spec_name =
  hash[test_spec.name] = pod_targets.reject { |t| test_dependencies[t.root_spec].nil? }
end
```

now both branches are using `filter_dependencies`, but is that valid? tests were passing but not sure if there's an edge case not covered by the tests

File: lib/cocoapods/installer/analyzer/target_generator.rb [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+            target_inspection = target_inspections[target_definition]
+            raise "missing inspection: #{target_definition.name}" unless target_inspection
+            user_project = target_inspection.project
+            client_root = user_project.path.dirname.realpath
+            user_target_uuids = target_inspection.project_target_uuids
+            user_build_configurations = target_inspection.build_configurations
+            archs = target_requires_64_bit ? ['$(ARCHS_STANDARD_64_BIT)'] : target_inspection.archs
+          else
+            user_project = nil
+            client_root = Config.instance.installation_root
```
Comment (amorde) [inline lib/cocoapods/installer/analyzer/target_generator.rb]: not happy about this part, but trying to keep the number of arguments from growing to a paragraph

File: lib/cocoapods/installer/analyzer/target_inspector.rb [hunk 3]
```diff
@@ -95,13 +97,12 @@ def compute_project_path
...
-        #         the user project
+        # @param  [Array<PBXNativeTarget>] native_targets
+        #         all native targets in the user project
         #
-        # @return [Array<PBXNativeTarget>]
+        # @return [Array<PBXNativeTarget>] the targets to be integrated
         #
-        def compute_targets(user_project)
-          native_targets = user_project.native_targets
+        def compute_targets(native_targets)
```
Comment (amorde) [inline lib/cocoapods/installer/analyzer/target_inspector.rb]: changed to accept just the array of native targets, since that's all it was using the project for

File: lib/cocoapods/resolver.rb [hunk 4]
```diff
@@ -133,12 +133,12 @@ def resolve
     #
     def resolver_specs_by_target
       @resolver_specs_by_target ||= {}.tap do |resolver_specs_by_target|
-        @podfile_dependency_cache.target_definition_list.each do |target|
+        @podfile_dependency_cache.target_definition_list.each do |target_definition|
```
Comment (amorde) [inline lib/cocoapods/resolver.rb]: renamed for clarity, no other logic changes here

**Comment by amorde** [comment]
not sure what's wrong with the build, I don't think the failures are related to this

**Comment by segiddins** [comment]
Restarted build

File: lib/cocoapods/installer/analyzer.rb [hunk 5]
```diff
@@ -196,6 +193,8 @@ def update_mode
         end
       end
 
+      # @return [Array<Pod::Dependency>]
```
Comment (dnkoutso) [inline lib/cocoapods/installer/analyzer.rb]: nit add more context here?

File: lib/cocoapods/installer/analyzer/target_generator.rb [hunk 6]
```diff
@@ -0,0 +1,397 @@
+module Pod
+  class Installer
+    class Analyzer
+      class TargetGenerator
+        # @return [String] The version of iOS which requires binaries with only 64-bit architectures
+        #
+        IOS_64_BIT_ONLY_VERSION = Version.new('11.0')
```
Comment (dnkoutso) [inline lib/cocoapods/installer/analyzer/target_generator.rb]: `.freeze`?

Comment (amorde) [reply lib/cocoapods/installer/analyzer/target_generator.rb]: ah, now I remember why I didn't do this - comparing versions with `>=` triggers a lazy initializer on `_segments` inside `Version`, which is a modification. so basically comparing one version with another modifies the version instance :/

**Comment by amorde** [comment]
looks like something broke after a rebase. will look into it at some point

File: lib/cocoapods/installer/analyzer/target_generator.rb [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+              test_dependencies = dependencies_for_specs([test_spec], target.platform, all_specs).group_by(&:root)
+              test_dependencies.delete_if { |k| dependencies.key? k }
+              hash[test_spec.name] = filter_dependencies(test_dependencies, pod_targets_by_name, target)
+            end
+          end
+        end
+
+        #-----------------------------------------------------------------------#
+
+        class << self
```
Comment (segiddins) [inline lib/cocoapods/installer/analyzer/target_generator.rb]: please be consistent, and either use `class << self` or `def self.`, but not both

File: lib/cocoapods/installer/analyzer/target_generator.rb (hunk 1 above)
Comment (amorde) [reply lib/cocoapods/installer/analyzer/target_generator.rb]: @segiddins this is the only thing that could potential affect functionality as far as I can tell

File: lib/cocoapods/installer/analyzer/target_generator.rb (hunk 7 above)
Comment (amorde) [reply lib/cocoapods/installer/analyzer/target_generator.rb]: I was trying to make this section private class methods, but it seems that private class methods aren't really a thing unless you use `class << self` is that correct? should I use `class << self` for the whole class then?

Comment (segiddins) [reply lib/cocoapods/installer/analyzer/target_generator.rb]: `private_class_ method :foo` is also a thing. I think we should just be consistent within a single class

**Comment by dnkoutso** [comment]
For 1.7.0

**Comment by amorde** [comment]
Going to close this and maybe revisit it in another PR.
